File: 3741_Minimum_Distance_Between_Three_Equal_Elements_II/solution.py
class Solution:
    def minimumDistance(self, a: List[int]) -> int:
        n, m = len(a), max(a)
        p = [(-1, -1) for _ in range(m + 1)]
        r = 1 << 32
        for i, x in enumerate(a):
            if p[x][1] != -1:
                r = min(r, (i - p[x][1]) << 1)
            p[x] = (i, p[x][0])
        return -1 if r == 1 << 32 else r

File: 3741_Minimum_Distance_Between_Three_Equal_Elements_II/solution.c
int minimumDistance(int* a, int n) {
    int m = 0;
    for (int i = 0; i < n; i++) if (a[i] > m) m = a[i];
    int p[m + 1][2];
    for (int i = 0; i <= m; i++) p[i][0] = p[i][1] = -1;
    int r = 1 << 30;
    for (int i = 0; i < n; i++) {
        int x = a[i];
        if (p[x][1] != -1) {
            int d = (i - p[x][1]) << 1;
            if (d < r) r = d;
        }
        int t = p[x][0];
        p[x][0] = i;
        p[x][1] = t;
    }
    return (r == (1 << 30)) ? -1 : r;
}
